dart/common/Addon.h

```diff
@@ -0,0 +1,304 @@
+
+  /// Use the StateMixer class to easily create a State extension from an
+  /// existing class or struct.
+  template <class Mixin>
+  using StateMixer = ExtensibleMixer<State, Mixin>;
+
+  /// If your Addon has a Properties class, then it should inherit this
+  /// Addon::Properties class. This allows us to safely serialize, store, and
+  /// clone the properties of arbitrary Addon extensions. If your Addon has no
+  /// properties, then you do not have to worry about extending this class,
+  /// because Addon::getProperties() will simply return a nullptr by default,
+  /// which is taken to indicate that it has no properties.
+  ///
+  /// The distinction between the State class and the Properties class is that
+  /// State will get stored in AddonManager::State whereas Properties will get
+  /// stored in AddonManager::Properties. Typically Properties are values that
+  /// only change rarely if ever, whereas State contains values that might
+  /// change as often as every time step.
+  class Properties : public Extensible<Properties> { };
+
+  /// Use the PropertiesMixer class to easily create a Properties extension
+  /// from an existing class or struct.
+  template <class Mixin>
+  using PropertiesMixer = ExtensibleMixer<Properties, Mixin>;
+
+  /// Virtual destructor
+  virtual ~Addon() = default;
+
+  /// Clone this Addon into a new manager
+  virtual std::unique_ptr<Addon> cloneAddon(AddonManager* newManager) const = 0;
+
+  /// Set the State of this Addon. By default, this does nothing.
+  virtual void setAddonState(const State& otherState);
+
+  /// Get the State of this Addon. By default, this returns a nullptr which
+  /// implies that the Addon is stateless.
+  virtual const State* getAddonState() const;
+
+  /// Set the Properties of this Addon. By default, this does nothing.
+  virtual void setAddonProperties(const Properties& someProperties);
+
+  /// Get the Properties of this Addon. By default, this returns a nullptr
+  /// which implies that the Addon has no properties.
+  virtual const Properties* getAddonProperties() const;
+
+  /// This function will be called if the user is attempting to delete the Addon
+  /// but not immediately replacing it with another Addon of the same type. The
+  /// incoming argument will point to the AddonManager that had been holding
+  /// this Addon.
+  ///
+  /// If your Addon is mandatory for the AddonManager type that is passed in
+  /// here, then you should perform error handling in this function, and you
+  /// should return false to indicate that the operation is not permitted. If
+  /// you return false, then the Addon will NOT be removed from its Manager.
+  ///
+  /// By default, this simply returns true.
+  virtual bool isOptional(AddonManager* oldManager);
+
+protected:
+
+  /// Constructor
+  ///
+  // We require the AddonManager argument in this constructor to make it clear
+  // to extensions that they must have an AddonManager argument in their
+  // constructors.
```

Comment:
Yeah, the message is useful for the developers, but, as you mentioned, the users would like to know the information if they want to create their own addons.